tainting: Assume sources and sanitizers are pure

IagoAbal opened the pull request:
Pure in the sense of having no side-effects.

In 0.94.0 we made it so when a pattern-source matched a variable exactly
this was understood as that variable being tainted by side-effect, e.g.
given `tainted(x)` we would taint `x` by side-effect. Same was done for
sanitizers.

This worked well initially, or so we thought, until in 0.113.0 we added
field-sensitivity, and in subsequent versions we applied sources and
sanitizers by side-effect on more general l-values. It was then that we
started to see regressions that were fairly unintuitive for users (see
added tests).

We now correct the situation by adding the `by-side-effect` option to
sources and sanitizers, and requiring this option to be explicitly enabled
in order to apply the source or the sanitizer by side-effect.

Requires https://github.com/returntocorp/semgrep-rules/pull/2577

Fixes: https://github.com/returntocorp/semgrep/commit/653db7dfa787b2e6057ee65dbf1ddb7c16b5bea0 ("tainting: Allow variables to be taint sources by side-effect (https://github.com/returntocorp/semgrep/pull/5240)")
Fixes: https://github.com/returntocorp/semgrep/commit/4ced4711358a6f19da87b59d9e7b3e69070bcdd1 ("tainting: Allow variables to be sanitized by side-effect (https://github.com/returntocorp/semgrep/pull/5266)")

Closes [PA-1629](https://linear.app/r2c/issue/PA-1629)
Closes [PA-2229](https://linear.app/r2c/issue/PA-2229)

Reviewed-by: Emma Jin

test plan:
make test # added three tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

emjin commented on semgrep-core/src/parsing/Parse_rule.ml:
```diff
@@ -1103,13 +1106,17 @@ let parse_taint_sanitizer ~(is_old : bool) env (key : key) (value : G.expr) =
   in
   let sanitizer_dict = yaml_to_dict env key value in
+  let sanitizer_mutate =
+    take_opt sanitizer_dict env parse_bool "mutate"
```
I'm assuming we're changing this from `"mutate"` to something else

IagoAbal replied:
Yes, finally it's going to be `taint-by-side-effect`.

emjin commented on semgrep-core/src/tainting/Dataflow_tainting.ml:
```diff
@@ -212,6 +212,11 @@ let merge_source_sink_mvars env source_mvars sink_mvars =
     (* The union of both sets, but taking the sink mvars in case of collision. *)
     sink_biased_union_mvars source_mvars sink_mvars
 
+let partition_mutable_sources sources_matches =
```
Personally I think `mutating_sources` is better? Mutable to me sounds like we're trying to identify if the source is mutable (e.g. a list in Python)

IagoAbal replied:
Agree, I will rename it.

emjin reviewed (approved):
In the changelog, it might be nice to give some examples of this unintuitive behavior to help people understand what this change entails. I don't think it's really easy to understand from a summary

IagoAbal committed:
tainting: Assume sources and sanitizers are pure

Pure in the sense of having no side-effects.

In 0.94.0 we made it so when a pattern-source matched a variable exactly
this was understood as that variable being tainted by side-effect, e.g.
given `tainted(x)` we would taint `x` by side-effect. Same was done for
sanitizers.

This worked well initially, or so we thought, until in 0.113.0 we added
field-sensitivity, and in subsequent versions we applied sources and
sanitizers by side-effect on more general l-values. It was then that we
started to see regressions that were fairly unintuitive for users (see
added tests).

We now correct the situation by adding the `by-side-effect` option to
sources and sanitizers, and requiring this option to be explicitly enabled
in order to apply the source or the sanitizer by side-effect.

Requires returntocorp/semgrep-rules#2577

Fixes: 653db7dfa78 ("tainting: Allow variables to be taint sources by side-effect (#5240)")
Fixes: 4ced4711358 ("tainting: Allow variables to be sanitized by side-effect (#5266)")

Closes PA-1629
Closes PA-2229

Reviewed-by: Emma Jin

test plan:
make test # added three tests
